core/src/main/scala/filodb.core/memstore/PartKeyIndex.scala

```diff
@@ -1,12 +1,228 @@
+        logger.warn(s"Exception while deleting directory for dataset=$ref, shard=$shardNum " +
+          s"and index directory=$indexDiskLocation with stack trace", t)
+        notifyLifecycleListener(IndexState.TriggerRebuild, System.currentTimeMillis)
+        throw new IllegalStateException("Unable to clean up index directory", t)
+    }
+  }
+  //else {
+  // TODO here we assume there is non-empty index which we need to validate
+  //}
+
+  protected def loadIndexData[T](ctor: () => T): T = try {
+    ctor()
+  } catch {
+    case e: Exception =>
+      // If an exception is thrown here there is something wrong with the index or the directory
+      // We will attempt once by cleaning the directory and try instantiating the index again
+      logger.warn(s"Index for dataset:${ref.dataset} and shard: $shardNum possibly corrupt," +
+        s"index directory will be cleaned up and index rebuilt", e)
+      Utils.deleteRecursively(indexDiskLocation.toFile) match {
+        case Success(_)       => // Notify the handler that the directory is now empty
+          logger.info(s"Cleaned directory for dataset=$ref," +
+            s"shard=$shardNum and index directory=$indexDiskLocation")
+          notifyLifecycleListener(IndexState.Empty, System.currentTimeMillis)
+        case Failure(t)       => logger.warn(s"Exception while deleting directory for dataset=$ref," +
+          s"shard=$shardNum and index directory=$indexDiskLocation with stack trace", t)
+          // If we still see failure, set the TriggerRebuild and rethrow the exception
+          notifyLifecycleListener(IndexState.TriggerRebuild, System.currentTimeMillis)
+          throw new IllegalStateException("Unable to clean up index directory", t)
+      }
+      // Retry again after cleaning up the index directory, if it fails again, something needs to be looked into.
+      ctor()
+  }
+
+
+  def notifyLifecycleListener(state: IndexState.Value, time: Long): Unit =
+    lifecycleManager.foreach(_.updateState(ref, shardNum, state, time))
+
+
+
+  protected def partKeyString(docId: String,
+                            partKeyOnHeapBytes: Array[Byte],
+                            partKeyBytesRefOffset: Int = 0): String = {
+    val partHash = schema.binSchema.partitionHash(partKeyOnHeapBytes, bytesRefToUnsafeOffset(partKeyBytesRefOffset))
+    //scalastyle:off
+    s"shard=$shardNum partId=$docId partHash=$partHash [${
+      TimeSeriesPartition
+        .partKeyString(schema, partKeyOnHeapBytes, bytesRefToUnsafeOffset(partKeyBytesRefOffset))
+    }]"
+    //scalastyle:on
+  }
+
+  private val utf8ToStrCache = concurrentCache[UTF8Str, String](PartKeyLuceneIndex.MAX_STR_INTERN_ENTRIES)
+
+  /**
+   * Map of partKey column to the logic for indexing the column (aka Indexer).
+   * Optimization to avoid match logic while iterating through each column of the partKey
+   */
+  protected final val indexers = schema.columns.zipWithIndex.map { case (c, pos) =>
```

Comment:
This is mostly moved as is without changes.  The calls to Lucene documents to add fields are replaced with addIndexedField and addIndexedMapField methods to allow override.